### PR TITLE
Feature/coloring: Four colorings for Zephyr and Pegasus, and a two coloring for Chimera

### PR DESCRIPTION
**Pull request by jackraymond**
Colorings of minimal chromatic number on Chimera, Pegasus and Zephyr graphs in the form of a map using the standard coordinate schemes.

### CODE REVIEW

**Comment by kevinchern** [comment]
@arcondello  I review this

File: tests/test_regular_coloring.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            for topology_shape in topology_shapes:
+                G = graph(*topology_shape, coordinates=True)
+                col_dict = {q: color(q) for q in G.nodes}
+                self.assertTrue(np.all(np.unique(list(col_dict.values())) ==
+                                       np.arange(num_colors)))
+                self.assertTrue(all(col_dict[q1] != col_dict[q2]
+                                    for q1, q2 in G.edges),
+                                f'{topology_type}[{topology_shape}]')
+
+    def test_non_default_schemes(self):
```
Comment (kevinchern) [inline tests/test_regular_coloring.py:52]: remove this test; see above comment using `lambda`

File: tests/test_regular_coloring.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+import dwave_networkx as dnx
+import numpy as np
+
+class TestRegularColoring(unittest.TestCase):
+
+    def test_valid(self):
+        test_cases = {'chimera': [(3,3), (4,4)],
+                      'pegasus': [(2,), (4,)],
+                      'zephyr': [(1,2), (3,4)]}
```
Comment (kevinchern) [inline tests/test_regular_coloring.py:25]: add tests for both zephyr schemes by, for example
```
...

'zephyr-scheme-0': [...],
'zephyr-scheme-1': [...],

...

color = lambda q: dnx.zephyr_color(q, scheme=1)
```

**Review by kevinchern — CHANGES_REQUESTED** [review]

File: dwave_networkx/generators/pegasus.py [hunk 3]
```diff
@@ -1303,3 +1304,28 @@ def relabel(u,w,k,z):
...
+    -------
+        color : int
+            Colors 0, 1, 2 or 3
+    Examples
+    ========
+    A mapping of every qubit (default integer labels) in the Pegasus[m]
+    graph to one of 4 colors
+    >>> m = 2
+    >>> G = dnx.pegasus_graph(m)
+    >>> colors = {q: dnx.pegasus_four_color(dnx.pegasus_coordinates(m).linear_to_zephyr(q)) for q in G.nodes()}    # doctest: +SKIP
```
[the remark this answers is not in the document]
Comment (jackraymond) [reply dwave_networkx/generators/pegasus.py]: Per offline suggestion, I added coordinates=True to simplify.

File: tests/test_regular_coloring.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                graph = dnx.chimera_graph
+                color = dnx.chimera_two_color
+                num_colors = 2
+            else:
+                raise ValueError('unrecognized topology')
+            
+            for topology_shape in topology_shapes:
+                G = graph(*topology_shape, coordinates=True)
+                col_dict = {q: color(q) for q in G.nodes}
+                self.assertTrue(np.all(np.unique(list(col_dict.values())) ==
```
Comment (kevinchern) [inline tests/test_regular_coloring.py:51]:
```suggestion
                self.assertSetEqual(set(col_dict.values()), set(range(num_colors)))
```

File: tests/test_regular_coloring.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                                       np.arange(num_colors)))
```
Comment (kevinchern) [inline tests/test_regular_coloring.py:52]:
```suggestion
```

**Review by kevinchern — APPROVED** [review]
can use set equality instead otherwise LGTM. passing the ball to poseidon/neptune (@arcondello / @randomir)

File: tests/test_regular_coloring.py (hunk 4 above)
Comment (kevinchern) [reply tests/test_regular_coloring.py:51]: might go past the pep8 linewidth limit

Comment (randomir) [reply tests/test_regular_coloring.py:51]: nit: `assertEqual` already uses `assertSetEqual` by default if sets are compared, so no need to be specific.

nit 2: it's enough for one of the values to be set-like (have `.difference()` method), so no need to actually convert both to a set.

File: tests/test_regular_coloring.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# Copyright 2024 D-Wave Systems Inc.
```
Comment (randomir) [inline tests/test_regular_coloring.py:1]: New (c) should be used in any new code:
```suggestion
# Copyright 2024 D-Wave Inc.
```

File: tests/test_regular_coloring.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import unittest
+
+import dwave_networkx as dnx
+import numpy as np
```
Comment (randomir) [inline tests/test_regular_coloring.py:18]: nit: [pep8 order](https://peps.python.org/pep-0008/#imports) is stdlib, 3rd-party, local

File: dwave_networkx/generators/chimera.py [hunk 8]
```diff
@@ -866,3 +867,29 @@ def chimera_torus(m, n=None, t=None, node_list=None, edge_list=None):
...
+        color : int
+            Colors 0, 1, 2 or 3
+
+    Examples
+    ========
+    A mapping of every qubit (default integer labels) in the Chimera[m, t]
+    graph to one of 2 colors
+    >>> m = 2
+    >>> G = dnx.chimera_graph(m, coordinates=True)
+    >>> colors = {q: dnx.chimera_two_color(q) for q in G.nodes()}    # doctest: +SKIP
```
Comment (randomir) [inline dwave_networkx/generators/chimera.py:891]: Why skip? (here and below)

File: tests/test_regular_coloring.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+class TestRegularColoring(unittest.TestCase):
+
+    def test_valid(self):
+        test_cases = {'chimera': [(3,3), (4,4)],
+                      'pegasus': [(2,), (4,)],
+                      'zephyr-0': [(1,2), (3,4)],
+                      'zephyr-1': [(1,3), (3,2)]}
+
+        for topology_type, topology_shapes in test_cases.items():
```
Comment (randomir) [inline tests/test_regular_coloring.py:28]: Good candidate for `parameterized`, if you feel like simplifying. For an example, you can check [here](https://github.com/dwavesystems/dwave-cloud-client/blob/b2432fe441d304162ee2f3bac6249241366ece91/tests/auth/test_server.py#L32-L42).

**Review by randomir — APPROVED** [review]
LGTM.

Happy to merge, or can wait if you want to address a few minor comments I made below.

File: dwave_networkx/generators/chimera.py (hunk 8 above)
Comment (jackraymond) [reply dwave_networkx/generators/chimera.py:891]: Lazy habit

File: tests/test_regular_coloring.py (hunk 9 above)
Comment (jackraymond) [reply tests/test_regular_coloring.py:28]: Next pull request

File: tests/test_regular_coloring.py (hunk 4 above)
Comment (jackraymond) [reply tests/test_regular_coloring.py:51]: range() has no .difference() method, so is required this case.

Comment (randomir) [reply tests/test_regular_coloring.py]: Right, types have to match as well.

**Pull request merged**